build: split depends Qt into native and target builds

fanquake opened the pull request:
This splits our Qt build in depends into two parts. The first builds `qmake`, qt tools (`uic`, `moc`, `rcc` etc) as well as translation tools (`lconvert`, `lupdate` etc). The second builds the libs we want for `bitcoin-qt`.

Splitting the build in this way has a few advantages. For example, we can enable the xml module while building the tools, and ideally fix issues like #14648 or #18536 without needing changes like #21589 or #21420, or having to build it alongside our libs.

This has been tested a little bit. GUIX builds are working for macOS and Windows.

Some testing still needs to be done in regards to passing compiler flags, DEBUG=1 builds etc. No doubt some improvements can still be made to what is currently here.

Fixes #18536.

fanquake committed:
build: split qt into native and host builds

hebasto committed:
build, qt: Fix wrong cross-compiling detection on macOS

hebasto commented:
Concept ACK. The same idea (a separated making of tools) is used for other dependencies.

Need more investigation to find out if this approach is compatible with adding additional Qt modules.

hebasto commented:
I think the reason of the failure on Centos 8 is wrong architecture the `qmake` is configured for -- `x86_64` instead of `i386`.

hebasto commented:
I'm able to `make -C depends qt_configured HOST=i686-pc-linux-gnu` with the correct architecture detection:
```
...
Configure summary:

Building on: linux-g++ (x86_64, CPU features: mmx sse sse2)
Building for: linux-g++-32 (i386, CPU features: <none>)
...
```
with the following patch:
```diff
--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -111,10 +111,6 @@ $(package)_config_opts += -no-feature-vulkan
 $(package)_config_opts += -no-feature-wizard
 $(package)_config_opts += -no-feature-xml
 
-$(package)_config_opts += "QMAKE_CFLAGS = '$($(package)_cflags) $($(package)_cppflags)'"
-$(package)_config_opts += "QMAKE_CXXFLAGS = '$($(package)_cflags) $($(package)_cppflags)'"
-$(package)_config_opts += "QMAKE_LFLAGS = '$($(package)_ldflags)'"
-
 $(package)_config_opts_darwin = -no-feature-corewlan
 $(package)_config_opts_darwin += QMAKE_MACOSX_DEPLOYMENT_TARGET=$(OSX_MIN_VERSION)
 
@@ -217,6 +213,9 @@ define $(package)_preprocess_cmds
   cp -f $($(package)_patch_dir)/mac-qmake.conf qtbase/mkspecs/macx-clang-linux/qmake.conf && \
   cp -r qtbase/mkspecs/linux-arm-gnueabi-g++ qtbase/mkspecs/bitcoin-linux-g++ && \
   sed -i.old "s/arm-linux-gnueabi-/$(host)-/g" qtbase/mkspecs/bitcoin-linux-g++/qmake.conf && \
+  echo "!host_build: QMAKE_CFLAGS     += $($(package)_cflags) $($(package)_cppflags)" >> qtbase/mkspecs/common/gcc-base.conf && \
+  echo "!host_build: QMAKE_CXXFLAGS   += $($(package)_cxxflags) $($(package)_cppflags)" >> qtbase/mkspecs/common/gcc-base.conf && \
+  echo "!host_build: QMAKE_LFLAGS     += $($(package)_ldflags)" >> qtbase/mkspecs/common/gcc-base.conf && \
   sed -i.old "s|QMAKE_CC                = \$$$$\$$$${CROSS_COMPILE}clang|QMAKE_CC                = $($(package)_cc)|" qtbase/mkspecs/common/clang.conf && \
   sed -i.old "s|QMAKE_CXX               = \$$$$\$$$${CROSS_COMPILE}clang++|QMAKE_CXX               = $($(package)_cxx)|" qtbase/mkspecs/common/clang.conf && \
   sed -i.old "s/LIBRARY_PATH/(CROSS_)?\0/g" qtbase/mkspecs/features/toolchain.prf
```

hebasto commented:
The branch with applied [fixup](https://github.com/bitcoin/bitcoin/pull/22142#issuecomment-856926951) is [here](https://github.com/hebasto/bitcoin/commits/210608-qtbase-5).
CI -- https://cirrus-ci.com/build/4834341327994880

hebasto commented:
> The branch with applied [fixup](https://github.com/bitcoin/bitcoin/pull/22142#issuecomment-856926951) is [here](https://github.com/hebasto/bitcoin/commits/210608-qtbase-5).

Tested on Armbian 21.02.3 -- #18536 is fixed, and #21589 could be dropped.

hebasto commented on depends/packages/native_qtbase.mk:
```diff
@@ -0,0 +1,193 @@
+$(package)_config_opts += -c++std c++1z
+$(package)_config_opts += -confirm-license
+$(package)_config_opts += -hostprefix $(build_prefix)
+$(package)_config_opts += -hostdatadir $(build_prefix)/share
```
This line seems unneeded. Verified with 9046de8a4cbc3899fed9eae084115f423e7ac5bd (from #21995).

fanquake replied:
Addressed in next push.

hebasto commented on depends/packages/native_qtbase.mk:
```diff
@@ -0,0 +1,193 @@
+
+define $(package)_extract_cmds
+  mkdir -p $($(package)_extract_dir) && \
+  echo "$($(package)_sha256_hash)  $($(package)_source)" > $($(package)_extract_dir)/.$($(package)_file_name).hash && \
+  echo "$($(package)_qttools_sha256_hash)  $($(package)_source_dir)/$($(package)_qttools_file_name)" >> $($(package)_extract_dir)/.$($(package)_file_name).hash && \
```
style nit, extra spaces:
```suggestion
  echo "$($(package)_sha256_hash) $($(package)_source)" > $($(package)_extract_dir)/.$($(package)_file_name).hash && \
  echo "$($(package)_qttools_sha256_hash) $($(package)_source_dir)/$($(package)_qttools_file_name)" >> $($(package)_extract_dir)/.$($(package)_file_name).hash && \
```

fanquake replied:
Addressed in next push.

hebasto commented:
The `linguist` stuff could be simplified:
```diff
--- a/depends/packages/native_qtbase.mk
+++ b/depends/packages/native_qtbase.mk
@@ -162,9 +162,7 @@ define $(package)_config_cmds
   qtbase/bin/qmake -o qtbase/src/tools/qfloat16-tables/Makefile qtbase/src/tools/qfloat16-tables/qfloat16-tables.pro && \
   qtbase/bin/qmake -o qtbase/src/tools/qvkgen/Makefile qtbase/src/tools/qvkgen/qvkgen.pro && \
   qtbase/bin/qmake -o qtbase/qmake/Makefile qtbase/qmake/qmake-aux.pro && \
-  qtbase/bin/qmake -o qttools/src/linguist/lrelease/Makefile qttools/src/linguist/lrelease/lrelease.pro && \
-  qtbase/bin/qmake -o qttools/src/linguist/lupdate/Makefile qttools/src/linguist/lupdate/lupdate.pro && \
-  qtbase/bin/qmake -o qttools/src/linguist/lconvert/Makefile qttools/src/linguist/lconvert/lconvert.pro
+  qtbase/bin/qmake -o qttools/src/linguist/Makefile qttools/src/linguist/linguist.pro
 endef
 
 define $(package)_build_cmds
@@ -174,9 +172,7 @@ define $(package)_build_cmds
   $(MAKE) -C qtbase/src/tools/qvkgen && \
   $(MAKE) -C qtbase/src/tools/rcc && \
   $(MAKE) -C qtbase/src/tools/uic && \
-  $(MAKE) -C qttools/src/linguist/lrelease && \
-  $(MAKE) -C qttools/src/linguist/lupdate && \
-  $(MAKE) -C qttools/src/linguist/lconvert
+  $(MAKE) -C qttools/src/linguist
 endef
 
 define $(package)_stage_cmds
```

Verified with 9046de8a4cbc3899fed9eae084115f423e7ac5bd (from #21995).

fanquake commented:
> I'm able to make -C depends qt_configured HOST=i686-pc-linux-gnu with the correct architecture detection:
> The linguist stuff could be simplified:

Thanks, I've addressed both of these points. Also addressed the comments you left inline and rebased for #22174. I think we can merge #22186 before this. I'm not really happy with how qt patches end up duplicated, but we should be able to fix that.

fanquake commented on depends/packages/qt.mk:
```diff
@@ -179,48 +167,36 @@ endef
 # way we can generically support hosts like riscv64-linux-gnu, which Qt doesn't
 # ship a mkspec for. See it's usage in config_opts_* above.
 #
-# 5. Put our C, CXX and LD FLAGS into gcc-base.conf. Only used for non-host builds.
```
#5 needs to be restored here.

hebasto commented:
> Splitting the build in this way has a few advantages. For example, we can enable the xml module while building the tools, and ideally fix issues like #14648 or #18536 without needing changes like #21589 or #21420, or having to build it alongside our libs.

It is true that avoiding intrusive fixes to the build system--i.e., patching Qt code base--makes it more maintainable and more comprehensible. That is why Concept ACKed this PR.

But considering different approaches, I still prefer the #20641. It is better than this one for the following reasons:
- no `qmake` building duplication
- no patch duplication
- using Qt "blessed" approach, that mimics building from the single source tree that includes all modules
- compatibility with possible including other modules in the future (e.g., QtQml for #16883, or QtMultimedia for Qr-code scanning)

fanquake commented:
> compatibility with possible including other modules in the future

Is this approach actually incompatible with doing that?

hebasto commented:
> > compatibility with possible including other modules in the future
> 
> Is this approach actually incompatible with doing that?

If new modules depends on core and tools are built in the `native_qtbase.mk` only -- that is ok.
But if they have mutual dependencies, this approach could fail (without additional hacks).

fanquake commented:
Closing this in favour of #20641 for now.
